// @state-space/hardhat/dist/testgen/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Network = exports.RequestStatus = exports.TestGenFixtureError = exports.TestGenError = exports.TestGenAuthRetryError = void 0;
class TestGenAuthRetryError extends Error {
}
exports.TestGenAuthRetryError = TestGenAuthRetryError;
class TestGenError extends Error {
}
exports.TestGenError = TestGenError;
class TestGenFixtureError extends Error {
    testSetName;
    sourceError;
    constructor(testSetName, sourceError) {
        super("while executing fixture" + (testSetName ? ' in "' + testSetName + '"' : "") + ": " + sourceError.message);
        this.testSetName = testSetName;
        this.sourceError = sourceError;
        this.cause = sourceError;
    }
}
exports.TestGenFixtureError = TestGenFixtureError;
var RequestStatus;
(function (RequestStatus) {
    RequestStatus["Completed"] = "Completed";
    RequestStatus["Config"] = "Config";
    RequestStatus["Created"] = "Created";
    RequestStatus["Pending"] = "Pending";
    RequestStatus["Post"] = "Post";
    RequestStatus["Processed"] = "Processed";
    RequestStatus["Started"] = "Started";
    RequestStatus["Waiting"] = "Waiting";
})(RequestStatus || (exports.RequestStatus = RequestStatus = {}));
var Network;
(function (Network) {
    Network["Mainnet"] = "Mainnet";
    Network["StateSpace"] = "StateSpace";
})(Network || (exports.Network = Network = {}));
